rfcs/0000-rlp-v2.md

```diff
@@ -0,0 +1,740 @@
+
+```yaml
+apiVersion: kuadrant.io/v2beta1
+kind: RateLimitPolicy
+metadata:
+  name: toystore-per-endpoint-per-user
+spec:
+  targetRef:
+    group: gateway.networking.k8s.io
+    kind: HTTPRoute
+    name: toystore
+  limits:
+    - name: readers
+      matches:
+        - path:
+            type: PathPrefix
+            value: "/toys"
+          method: GET
+      rates:
+        - limit: 50
+          duration: 1
+          unit: minute
+      counters:
+        - auth.identity.username
+      when:
+        - selector: auth.identity.group
+          operator: neq
+          value: admin
+
+    - name: writers
+      matches:
+        - path:
+            type: PathPrefix
+            value: "/toys"
+          method: POST
+      rates:
+        - limit: 5
+          duration: 1
+          unit: minute
+        - limit: 100
+          duration: 12
+          unit: hour
+      counters:
+        - auth.identity.username
+      when:
+        - selector: auth.identity.group
+          operator: neq
+          value: admin
+```
+
+##### Example 3. Route filtering using `when` conditions
+
+`when` conditions are preferably to be used for special cases of conditional filtering based on values other than attributes of the route that could otherwise be specified using the `matches` field. However, in some cases such as where the condition is not related to the HTTPRoute (e.g. see [Example 2](#example-2-specific-route-matching-rule-targeted-conditions-counter-qualifiers-and-multiple-rates) above; filterring based on hostname, etc) or a temporary measure while the target object misses the desired matching rule, the `when` conditions remains an option.
+
+In this example, a special limit with one rate limit of 150 rps is set for `GET` requests to the `/toys/special` path. Because the `toystore` HTTPRoute is unaltered, the special case is defined using the `when` conditions field.
+
+```yaml
+apiVersion: kuadrant.io/v2beta1
+kind: RateLimitPolicy
+metadata:
+  name: toystore-special-toys
+spec:
+  targetRef:
+    group: gateway.networking.k8s.io
+    kind: HTTPRoute
+    name: toystore
+  limits:
+    - name: read-special
+      matches:
+        - path:
+            type: PathPrefix
+            value: "/toys"
+          method: GET
+      rates:
+        - limit: 150
+          unit: second
+      when:
+        - selector: context.request.http.path
+          operator: eq
+          value: /toys/special
+```
+
+By using the route rule as-is, simply `GET /toys*` instead of more specific `GET /toys/special`, the rate limit filter (wasm filter) will continue invoking the rate limit service (Limitador) for all requests that match HTTP method equal to `GET` and path prefix equal to `/toys`, including requests to paths other than `/toys/special`, but the limit will only be enforced when the path fully matches `/toys/special`, as specified in the `when` condition.
```

Comment:
For the reasons for distinguishing between `matches` and `when`, please see https://github.com/Kuadrant/architecture/pull/8#discussion_r1118617726.

Another pointed raised here, about the wasm plugin in the data plane being a centralised decision point where all kinds of conditions could be evaluated before handling control to Limitador, I see as slightly out of the scope of this RFC.

The idea seems to root in the fact that, according to current implementation, all traffic has to pass through the wasm plugin, a component owned by Kuadrant. This is a side-effect we once absorbed from a decision made due to limitations of GW-API and the Istio resources used to inject the configuration in Envoy. In this case, it's arguably a "positive" side-effect that would allow us to treat all kinds of conditions as one. However, the fact that the wasm filter kicks in every time in itself was, IIRC, another "less positive" side effect of that decision.

If we want to embrace the current implementation based on the wasm filter for good, and break away from what otherwise could be based on plain Envoy's Rate Limit filter, where sometimes no Kuadrant component is involved in handling the request, I think it's a proposal worth discussing. I just didn't want it to be an enforcement imposed by this RFC.

Reply:
I somewhat agree with @eguzki here, but for a different reason, how this happens and whether we could do things differently (and more efficiently) within the Envoy's WASM filter is up to us and probably best left out (if only of this particular section) of the RFC.

Reply:
>  However, the fact that the wasm filter kicks in every time in itself was, IIRC, another "less positive" side effect of that decision.

Kuadrant's wasm filter is (yet) another envoy filter chain entry, like envoys rate limit filter. It will be kicked in anyway for each request, even if there is no HTTPRoute to wire with backend.